### PR TITLE
Set the right shape of selected_rows

### PR DESCRIPTION
**Pull request by chengduoZH**
The output shape of some operator is wrong when the input var's type is SelectedRows.
 
Some developers use [`GetInputDim`](https://github.com/PaddlePaddle/Paddle/blob/41e4f7ea0a6479e14f9ec7e15fcc358e306596bb/paddle/fluid/framework/shape_inference.cc#L25) to get the input var's shape, it is not right, because if the input var's type is SelectedRows, `GetInputDim` will return the completed dim of the SelectedRows in the running time, the code is [here](https://github.com/PaddlePaddle/Paddle/blob/41e4f7ea0a6479e14f9ec7e15fcc358e306596bb/paddle/fluid/framework/operator.cc#L602). But the [`GetCompleteDims `](https://github.com/PaddlePaddle/Paddle/blob/41e4f7ea0a6479e14f9ec7e15fcc358e306596bb/paddle/fluid/framework/selected_rows.h#L125-L129) is not the dim of the tensor of  the SelectedRows, it causes the dim of the output var is wrong.

### CODE REVIEW

File: python/paddle/fluid/clip.py [hunk 1]
```diff
@@ -264,15 +264,15 @@ def _process_context(self, context, param, grad):
...
         else:
             if not self.clip_norm == context[self.group_name + "_clip_value"]:
                 raise ValueError(
                     "All parameters' 'clip_norm' of a same group should be the same"
                 )
 
-        square = grad * grad
-        local_norm_var = layers.cast(layers.reduce_sum(input=square), 'float64')
+        square = layers.pow(grad, 2.0)
+        local_norm_var = layers.reduce_sum(input=square)
```
Comment (panyx0718) [inline python/paddle/fluid/clip.py]: float64 is no longer needed?

File: paddle/fluid/operators/lookup_table_op.cc [hunk 2]
```diff
@@ -137,6 +137,7 @@ class LookupTableOpGradVarTypeInference : public framework::VarTypeInference {
               << " is set to LoDTensor";
       block->Var(out_var_name)->SetType(framework::proto::VarType::LOD_TENSOR);
     }
+    block->Var(out_var_name)->SetDataType(block->Var("W")->GetDataType());
```
Comment (panyx0718) [inline paddle/fluid/operators/lookup_table_op.cc:140]: why this change?

File: paddle/fluid/operators/elementwise_op.h [hunk 3]
```diff
@@ -193,14 +194,12 @@ class ElementwiseOpExplicitGrad : public ElementwiseOpGrad {
...
-      auto out_dims = ctx->GetInputDim(framework::GradVarName("Out"));
-      ctx->SetOutputDim(x_grad_name, out_dims);
+      ctx->ShareDimAndLod(framework::GradVarName("Out"), x_grad_name);
     }
     auto y_grad_name = framework::GradVarName("Y");
     if (ctx->HasOutput(y_grad_name)) {
       PADDLE_ENFORCE(ctx->HasInput("Y"), "Input(Y) should not be null");
-      auto y_dims = ctx->GetInputDim("Y");
-      ctx->SetOutputDim(y_grad_name, y_dims);
+      ctx->ShareDimAndLod("Y", y_grad_name);
```
Comment (panyx0718) [inline paddle/fluid/operators/elementwise_op.h]: Why share lod?

File: paddle/fluid/operators/elementwise_op.h [hunk 4]
```diff
@@ -193,14 +194,12 @@ class ElementwiseOpExplicitGrad : public ElementwiseOpGrad {
 
     auto x_grad_name = framework::GradVarName("X");
     if (ctx->HasOutput(x_grad_name)) {
-      auto out_dims = ctx->GetInputDim(framework::GradVarName("Out"));
-      ctx->SetOutputDim(x_grad_name, out_dims);
+      ctx->ShareDimAndLod(framework::GradVarName("Out"), x_grad_name);
```
Comment (panyx0718) [inline paddle/fluid/operators/elementwise_op.h]: why share lod?

File: paddle/fluid/operators/elementwise_op.h [hunk 5]
```diff
@@ -157,10 +158,10 @@ class ElementwiseOpGrad : public framework::OperatorWithKernel {
     auto x_grad_name = framework::GradVarName("X");
     auto y_grad_name = framework::GradVarName("Y");
     if (ctx->HasOutput(x_grad_name)) {
-      ctx->SetOutputDim(x_grad_name, x_dims);
+      ctx->ShareDimAndLod("X", x_grad_name);
```
Comment (panyx0718) [inline paddle/fluid/operators/elementwise_op.h]: why share lod?

File: paddle/fluid/framework/operator.cc [hunk 6]
```diff
@@ -579,6 +579,39 @@ class RuntimeInferShapeContext : public InferShapeContext {
...
+      auto& in_sele_rows = in_var->Get<framework::SelectedRows>();
+      auto out_sele_rows = out_var->GetMutable<framework::SelectedRows>();
+      out_sele_rows->mutable_value()->set_layout(in_sele_rows.value().layout());
+      out_sele_rows->mutable_value()->Resize(in_sele_rows.value().dims());
+      out_sele_rows->set_rows(in_sele_rows.rows());
+      out_sele_rows->set_height(in_sele_rows.height());
+    } else if (in_var->IsType<framework::LoDTensor>()) {
+      auto& in_lod_tensor = in_var->Get<framework::LoDTensor>();
+      auto* out_lod_tensor = out_var->GetMutable<framework::LoDTensor>();
+      out_lod_tensor->Resize(in_lod_tensor.dims());
```
Comment (panyx0718) [inline paddle/fluid/framework/operator.cc:567]: why resize? not set dim?

File: paddle/fluid/operators/activation_op.cc [hunk 7]
```diff
@@ -91,12 +90,25 @@ class ActivationOp : public framework::OperatorWithKernel {
...
+  }
+};
+
 class ActivationOpGrad : public framework::OperatorWithKernel {
  public:
   using framework::OperatorWithKernel::OperatorWithKernel;
 
   void InferShape(framework::InferShapeContext* ctx) const override {
-    ctx->SetOutputDim(framework::GradVarName("X"), ctx->GetInputDim("Out"));
+    ctx->ShareDimAndLod("Out", framework::GradVarName("X"));
```
Comment (panyx0718) [inline paddle/fluid/operators/activation_op.cc]: why share lod

File: python/paddle/fluid/clip.py [hunk 8]
```diff
@@ -264,15 +264,15 @@ def _process_context(self, context, param, grad):
             context[self.group_name] = []
             context[self.group_name + "_clip_value"] = self.clip_norm
             context[self.group_name + "_clip"] = layers.fill_constant(
-                shape=[1], dtype="float32", value=self.clip_norm)
+                shape=[1], dtype=grad.dtype, value=self.clip_norm)
```
Comment (chengduoZH) [inline python/paddle/fluid/clip.py]: I think `dtype="float32"` is a hard code here, the type should be the same with grad, right? @ktlichkid

File: python/paddle/fluid/clip.py [hunk 9]
```diff
@@ -282,7 +282,6 @@ def _create_operators(self, param, grad):
         if group_scale_name not in self.context:
             group_norm_var = layers.sums(input=self.context[self.group_name])
             group_norm_var = layers.sqrt(x=group_norm_var)
-            group_norm_var = layers.cast(group_norm_var, 'float32')
```
Comment (chengduoZH) [inline python/paddle/fluid/clip.py]: This line is added by @ktlichkid in this PR https://github.com/PaddlePaddle/Paddle/pull/13516, I think this is because we didn't set the DataType in Activation_op, so the data type of `layers.sqrt(x=group_norm_var)` is BOOL, is it right? @ktlichkid
Now I have fixed it [here](https://github.com/PaddlePaddle/Paddle/pull/13723/files#diff-b6a56915c851e097c696f5ef8a5724cfR102).

File: python/paddle/fluid/clip.py (hunk 1 above)
Comment (chengduoZH) [reply python/paddle/fluid/clip.py]: I am not sure, the original code of line 275 is written by @ktlichkid, can you explain why use `float64` here?

File: paddle/fluid/operators/activation_op.cc (hunk 7 above)
Comment (chengduoZH) [reply paddle/fluid/operators/activation_op.cc]: We should share the lod in all operators because the lod is a part of LodTensor, if not, it may be wrong.

File: paddle/fluid/framework/op_desc.cc [hunk 10]
```diff
@@ -50,6 +50,28 @@ class CompileTimeInferShapeContext : public InferShapeContext {
   const std::vector<std::string> &Outputs(
       const std::string &name) const override;
 
+  void ShareDimAndLod(const std::string &in, const std::string &out,
```
Comment (dzhwinter) [inline paddle/fluid/framework/op_desc.cc]: Why we should combine the ShareDimAndLod into one API?  The i, j is for the dims, the lod do not need them at all.

File: paddle/fluid/framework/operator.cc [hunk 11]
```diff
@@ -579,6 +579,39 @@ class RuntimeInferShapeContext : public InferShapeContext {
...
+                   "The type of %s and %s is not the same.", output_n,
+                   GetDim(input_n));
+
+    if (in_var->IsType<framework::SelectedRows>()) {
+      auto& in_sele_rows = in_var->Get<framework::SelectedRows>();
+      auto out_sele_rows = out_var->GetMutable<framework::SelectedRows>();
+      out_sele_rows->mutable_value()->set_layout(in_sele_rows.value().layout());
+      out_sele_rows->mutable_value()->Resize(in_sele_rows.value().dims());
+      out_sele_rows->set_rows(in_sele_rows.rows());
+      out_sele_rows->set_height(in_sele_rows.height());
```
Comment (dzhwinter) [inline paddle/fluid/framework/operator.cc:563]: here also need set_layout.

**Review by dzhwinter — COMMENTED** [review]
The LoD always be the dangling part of the Tensor, maybe we need to re-cap the LoD cons and pros.

File: paddle/fluid/framework/op_desc.cc (hunk 10 above)
Comment (chengduoZH) [reply paddle/fluid/framework/op_desc.cc]: > Why we should combine the ShareDimAndLod into one API?

There are two main reasons that combining the ShareLod and ShareDim, 
1. In most operator, the output var's dim and lod is the same with one of the input vars, for example, activation op and elementwise_op.
2. most logic of  ShareDim and ShareLod are the same.

> The i, j is for the dims, the lod do not need them at all.

i and j indicate which argument of `in` and `out`, they are not only for the dim.

File: paddle/fluid/framework/operator.cc (hunk 11 above)
Comment (chengduoZH) [reply paddle/fluid/framework/operator.cc:563]: ~~Right!~~
line 598 has set the layout.

File: python/paddle/fluid/clip.py (hunk 1 above)
Comment (ktlichkid) [reply python/paddle/fluid/clip.py]: The computation here sums the square of gradients up globally, thus it may easily cause overflow. Float64 here is to prevent such overflows, and after the square root computation this value is again cast back to float32.

File: python/paddle/fluid/clip.py (hunk 9 above)
Comment (ktlichkid) [reply python/paddle/fluid/clip.py]: This is to cast temporary float64 (to prevent overflow) value back to float32.

File: paddle/fluid/operators/lookup_table_op.cc (hunk 2 above)
Comment (panyx0718) [reply paddle/fluid/operators/lookup_table_op.cc:140]: set data type in python to be consistent? (or all move to C++? which seems a lot of work)

File: python/paddle/fluid/clip.py (hunk 1 above)
Comment (panyx0718) [reply python/paddle/fluid/clip.py]: revert this change?

**Review by panyx0718 — APPROVED** [review]
